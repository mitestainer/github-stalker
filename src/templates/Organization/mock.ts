export const cardsMock = [
  {
    id: 1,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 2,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 3,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 4,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 5,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 6,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 7,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 8,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 9,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 10,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 11,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 12,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  },
  {
    id: 13,
    name: 'react',
    description:
      'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    language: 'javascript',
    stars: 123,
    url: 'https://github.com/facebook',
    updatedAt: '2021-11-14T14:03:02Z'
  }
]

export const mock = {
  handle: 'facebook',
  name: 'Meta',
  avatar: '/img/meta.png',
  description:
    'We are working to build community through open source technology. NB: members must have two-factor auth.',
  location: 'Menlo Park, California',
  website: 'https://opensource.fb.com',
  reposCount: 13
}
