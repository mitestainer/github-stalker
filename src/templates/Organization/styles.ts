import styled, { css } from 'styled-components'
import { Container as DefaultContainer } from 'components/Container'
import { Wrapper as BioCardWrapper } from 'components/BioCard/styles'
import { Button } from 'components/Button/styles'

export const Container = styled(DefaultContainer)`
  ${({ theme }) => css`
    margin: ${theme.gutter} auto;
    width: 100%;

    @media screen and (max-width: 768px) {
      margin: ${theme.gutter};
    }
  `}
`

export const Wrapper = styled.main`
  ${({ theme }) => css`
    height: 100%;
    display: flex;
    flex-direction: column;

    ${BioCardWrapper} {
      margin: ${theme.gutter} 0;
    }
  `}
`

export const Section = styled.section`
  display: flex;
  flex-direction: column;
  flex: 1 1;
`

export const Repositories = styled.div`
  ${({ theme }) => css`
    padding: ${theme.gutter};
    background-color: ${theme.colors.white};
    border-radius: ${theme.borderRadius};
    flex: 1 1;
    display: flex;
    flex-direction: column;
    row-gap: ${theme.gutter};

    > h2 {
      font-size: ${theme.font.size.medium};
    }

    ${Button} {
      background-color: ${theme.colors.lightgray};
      color: ${theme.colors.gray};
      width: 10rem;
      height: 4rem;
      margin: 0 auto;

      &:hover,
      &:disabled {
        opacity: 0.8;
      }
    }
  `}
`

type GridProps = {
  empty: boolean
}

export const Grid = styled.div<GridProps>`
  ${({ theme, empty }) => css`
    display: ${empty ? 'flex' : 'grid'};

    ${empty
      ? css`
          flex: 1 1;
          flex-direction: column;
          justify-content: center;
          align-items: center;
          row-gap: ${theme.gutter};
          color: ${theme.colors.gray};
        `
      : css`
          grid-template-columns: repeat(auto-fit, minmax(35rem, 1fr));
          gap: ${theme.gutter};
        `}

    @media screen and (max-width: 768px) {
      grid-template-columns: 1fr;
      grid-template-columns: repeat(auto-fit, minmax(25rem, 1fr));
    }
  `}
`
