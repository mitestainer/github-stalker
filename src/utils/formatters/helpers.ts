export const formatDate = (dateString: string) => {
  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(new Date(dateString))
}
