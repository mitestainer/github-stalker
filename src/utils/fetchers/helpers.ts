import { formatDate } from 'utils/formatters/helpers'

type RepoResponseProps = {
  id: number
  full_name: string
  description: string
  html_url: string
  watchers: number
  language: string
  updated_at: string
}

export const fetchRepos = async (handle: string, page: number) => {
  const response = await fetch(
    `https://api.github.com/orgs/${handle}/repos?sort=updated&direction=desc&page=${page}&per_page=12`
  )
  const data = await response.json()

  return data.map((repo: RepoResponseProps) => ({
    id: repo.id,
    name: repo.full_name,
    description: repo.description,
    url: repo.html_url,
    stars: repo.watchers,
    language: repo.language,
    updatedAt: formatDate(repo.updated_at)
  }))
}
