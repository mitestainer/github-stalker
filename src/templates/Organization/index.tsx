import { useEffect, useState } from 'react'
import Image from 'next/image'
import BioCard from 'components/BioCard'
import Button from 'components/Button'
import Empty from 'components/Empty'
import RepoCard, { RepoCardProps } from 'components/RepoCard'
import SearchBar from 'components/SearchBar'
import * as S from './styles'
import { fetchRepos } from '../../utils/fetchers/helpers'

type ReposProps = {
  id: number
} & RepoCardProps

export type OrganizationProps = {
  handle: string
  name: string
  description: string
  avatar: string
  location: string
  website: string
  reposCount: number
  repos: ReposProps[]
}

const OrganizationTemplate = ({
  handle,
  name,
  description,
  avatar,
  location,
  website,
  reposCount,
  repos
}: OrganizationProps) => {
  const [loading, setLoading] = useState(false)
  const [reposList, setReposList] = useState<ReposProps[]>([])
  const [nextPageIndex, setNextPageIndex] = useState(1)

  useEffect(() => {
    setReposList(repos)
  }, [repos])

  const pagesAmount = Math.ceil(reposCount / 12)

  const handleShowMore = async () => {
    setLoading(true)
    setNextPageIndex(reposList.length / 12 + 1)
    const newRepos = await fetchRepos(handle, nextPageIndex)
    setReposList([...reposList, ...newRepos])
    setLoading(false)
  }

  return (
    <S.Container>
      <S.Wrapper>
        <SearchBar withLogo />
        <S.Section>
          <BioCard
            name={name}
            description={description}
            avatar={avatar}
            location={location}
            website={website}
          />
          <S.Repositories>
            <h2>Repositories ({reposCount})</h2>
            <S.Grid empty={!reposCount}>
              {reposCount > 0 ? (
                reposList.map((card: ReposProps) => (
                  <RepoCard key={card.id} {...card} />
                ))
              ) : (
                <Empty />
              )}
            </S.Grid>
            {reposCount > 0 && nextPageIndex < pagesAmount ? (
              <Button disabled={loading} onClick={handleShowMore}>
                {loading ? (
                  <Image src="/img/loading.svg" height={20} width={20} />
                ) : (
                  'Load more'
                )}
              </Button>
            ) : null}
          </S.Repositories>
        </S.Section>
      </S.Wrapper>
    </S.Container>
  )
}

export default OrganizationTemplate
