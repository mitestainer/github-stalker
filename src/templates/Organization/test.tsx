import { render, screen } from 'utils/tests/helpers'

import Organization from '.'
import { mock, cardsMock } from './mock'

jest.mock('components/SearchBar', () => {
  return {
    __esModule: true,
    default: function () {
      return <div data-testid="Mock SearchBar"></div>
    }
  }
})

jest.mock('components/BioCard', () => {
  return {
    __esModule: true,
    default: function () {
      return <div data-testid="Mock BioCard"></div>
    }
  }
})

jest.mock('components/RepoCard', () => {
  return {
    __esModule: true,
    default: function () {
      return <div data-testid="Mock RepoCard"></div>
    }
  }
})

jest.mock('components/Empty', () => {
  return {
    __esModule: true,
    default: function () {
      return <div data-testid="Mock Empty"></div>
    }
  }
})

describe('<Organization />', () => {
  it('should render the organization page', () => {
    const { container } = render(<Organization {...mock} repos={cardsMock} />)

    expect(screen.getByTestId('Mock SearchBar')).toBeInTheDocument()
    expect(screen.getByTestId('Mock BioCard')).toBeInTheDocument()
    expect(
      screen.getByText(`Repositories (${mock.reposCount})`)
    ).toBeInTheDocument()
    expect(screen.getAllByTestId('Mock RepoCard')).toHaveLength(13)
    expect(
      screen.getByRole('button', { name: /load more/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the Empty component when there are no repos to show', () => {
    render(<Organization {...mock} repos={[]} reposCount={0} />)

    expect(screen.getByText('Repositories (0)')).toBeInTheDocument()
    expect(screen.getByTestId('Mock Empty')).toBeInTheDocument()
  })

  it('shouldn\'t render "Load more" button if there are less than 12 cards on screen', () => {
    render(
      <Organization
        {...mock}
        reposCount={3}
        repos={cardsMock.filter((item) => item.id < 4)}
      />
    )

    expect(screen.getAllByTestId('Mock RepoCard')).toHaveLength(3)
    expect(
      screen.queryByRole('button', { name: /load more/i })
    ).not.toBeInTheDocument()
  })
})
