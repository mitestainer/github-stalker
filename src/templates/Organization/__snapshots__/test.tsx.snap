// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Organization /> should render the organization page 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  border-radius: 1rem;
  padding: 1rem 1.5rem;
  cursor: pointer;
}

.c0 {
  max-width: 120rem;
  margin: 0 auto;
  background-color: #eff1fa;
  padding: 2rem;
  border-radius: 1rem;
}

.c1 {
  margin: 2rem auto;
  width: 100%;
}

.c2 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
}

.c4 {
  padding: 2rem;
  background-color: #fff;
  border-radius: 1rem;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 2rem;
}

.c4 > h2 {
  font-size: 1.6rem;
}

.c4 .c6 {
  background-color: #dadae5;
  color: #8383a6;
  width: 10rem;
  height: 4rem;
  margin: 0 auto;
}

.c4 .c6:hover,
.c4 .c6:disabled {
  opacity: 0.8;
}

.c5 {
  display: grid;
  grid-template-columns: repeat(auto-fit,minmax(35rem,1fr));
  gap: 2rem;
}

@media screen and (max-width:768px) {
  .c0 {
    max-width: 100%;
  }
}

@media screen and (max-width:768px) {
  .c1 {
    margin: 2rem;
  }
}

@media screen and (max-width:768px) {
  .c5 {
    grid-template-columns: 1fr;
    grid-template-columns: repeat(auto-fit,minmax(25rem,1fr));
  }
}

<div
  class="c0 c1"
>
  <main
    class="c2"
  >
    <div
      data-testid="Mock SearchBar"
    />
    <section
      class="c3"
    >
      <div
        data-testid="Mock BioCard"
      />
      <div
        class="c4"
      >
        <h2>
          Repositories (
          13
          )
        </h2>
        <div
          class="c5"
          data-cy="grid"
        >
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
          <div
            data-testid="Mock RepoCard"
          />
        </div>
        <button
          class="c6 c7"
        >
          Load more
        </button>
      </div>
    </section>
  </main>
</div>
`
